### PR TITLE
feat(composer): handle providers at packages.json root level

### LINKED ISSUE #6025: Support providers for composer/packagist
**What would you like Renovate to be able to do?**
Add support for `providers` from composer registries at root level. currently only `provider-includes` is supported which is used by [packagist itself](https://repo.packagist.org/packages.json).
https://github.com/renovatebot/renovate/blob/aeb6bfcd4dacaa7cfa8c57a1d0583cbe72ee405f/lib/datasource/packagist/index.ts#L63-L73

We have recently created our own [satis repository](https://github.com/composer/satis) which creates a json like this:
```json
{
  "packages": [],
  "providers-url": "/p/%package%$%hash%.json",
  "providers": {
    "abraham/twitteroauth": {
      "sha256": "37b194017aead928b4b6fb3676180d322134c301d9338ba13ccfd5274bba91b2"
    },
    "algo26-matthias/idna-convert": {
      "sha256": "129705c96b7565b2141870c59b2e98cdd8b29662e70df0a36cb423ab60b1199f"
    },
    "andreas-weber/php-junit-merge": {
      "sha256": "7b9fed898c2388edcb8fb22575719a926cb5270b9b78f4bd3f0262b93240762a"
    },
    "aws/aws-sdk-php": {
      "sha256": "684e9100020aff4fcbf728728e1f0890021483d33604f964064e89a5e5d5e171"
    }
  }
}
```

**Describe the solution you'd like**
Support the providers 

**Describe alternatives you've considered**
None

**Additional context**
This is our satis configuration (redacted)
```json
{
    "$schema": "https://raw.githubusercontent.com/composer/satis/master/res/satis-schema.json",
    "name": "company/satis",
    "description": "Satis Repository for Company",
    "homepage": "https://composer.company",
    "require-all": true,
    "require-dependencies": true,
    "archive": {
        "directory": "dist",
        "format": "zip",
        "prefix-url": "https://composer.company",
        "skip-dev": true
    },
    "providers": true,
    "pretty-print": false,
    "repositories-dep": [
        {
            "type": "composer",
            "url": "https://packagist.org"
        }
    ],
    "repositories": [
        {
            "name": "company/package",
            "type": "git",
            "url": "git@company:package.git"
        }
    ]
}
```

Like my previous feature requests i am willing to create an implementation

### PR DESCRIPTION
**Pull request by Tanuel**
Closes #6025 

i changed the test names to resemble the field-name in packages.json

There might be unintended/buggy behaviour when a package exists in multiple registries.
This can happen when using `"require-dependencies": true` in the satis configuration but not disabling packagist. But that is a topic for a different time

### CODE REVIEW

File: lib/datasource/packagist/index.ts [hunk 1]
```diff
@@ -38,6 +39,7 @@ interface RegistryFile {
 }
 interface RegistryMeta {
   files?: RegistryFile[];
+  providerPackages: Record<string, string>;
```
Comment (JamieMagee) [inline lib/datasource/packagist/index.ts:42]: This is now a required property?

Comment (Tanuel) [reply lib/datasource/packagist/index.ts:42]: The variable would be created anyways in here: https://github.com/renovatebot/renovate/pull/6028/files#diff-9d4e78c90551425d7672fa30aec4d667L180

I moved it to the registry meta so we can extract providers at root level here: https://github.com/renovatebot/renovate/pull/6028/files#diff-9d4e78c90551425d7672fa30aec4d667R53-R55

Previously, providers would only be supported from the `provider-includes` files (see https://repo.packagist.org/packages.json) but they can exist in packages.json themselves (see https://composer.typo3.org/packages.json for an example) which would previously go unnoticed

**Comment by Tanuel** [comment]
> Do you have a sample PR?

Sorry i have no idea how i would provide one

**Comment by JamieMagee** [comment]
> Sorry i have no idea how i would provide one

Create a small repository on GitHub, or wherever, and run this branch of Renovate against it.

File: lib/datasource/packagist/index.ts (hunk 1 above)
Comment (JamieMagee) [reply lib/datasource/packagist/index.ts:42]: I was more referring to the fact that to create an empty `providerPackages` for any `RegistryMeta` object. Is it necessary? Like here:

https://github.com/renovatebot/renovate/blob/65a8f33ad9ba797e5cdeaeac9259f6ae90fb923d/lib/datasource/packagist/index.ts#L53-L55

Comment (Tanuel) [reply lib/datasource/packagist/index.ts:42]: Alternative would be to create it only if necessary and later have a check and create it anyways. I just moved to creation to an earlier moment

**Comment by Tanuel** [comment]
@JamieMagee i ran this from my local machine, using the public typo3 repository https://github.com/Tanuel/renovate-packagist-providers-test/pull/2

**Comment by Tanuel** [comment]
Btw, running renovate from current master results in the following error:
```json
{
  "name": "renovate",
  "hostname": "tanuel",
  "pid": 18436,
  "level": 20,
  "logContext": "5cfHIJ3yu",
  "repository": "Tanuel/renovate-packagist-providers-test",
  "dependency": "typo3-ter/googleanalytics",
  "packageFile": "composer.json",
  "msg": "Failed to look up dependency typo3-ter/googleanalytics",
  "time": "2020-04-24T07:02:55.191Z",
  "v": 0
}
```

**Pull request merged**